### PR TITLE
Laboratory work #1, Yefimenko Natalya - 23FPL2

### PR DESCRIPTION
**Pull request by ShrubberyBush**

### CODE REVIEW

**Comment by ShrubberyBush** [comment]
a

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -21,8 +22,17 @@ def tokenize(text: str) -> list[str] | None:
     In case of corrupt input arguments, None is returned
     """
 
+    tokens = []
+    if isinstance(text, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:26]: Swap the condition. Make function return None instantly if type is wrong. Dont place all code in if statement.

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -21,8 +22,17 @@ def tokenize(text: str) -> list[str] | None:
...
+        for symb in text:
+            if symb.isalpha():
+                tokens.append(symb.lower())
+        return tokens
+    return None
+
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:35]: delete whitespace

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -34,9 +44,24 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list) or len(tokens) == 0 or None in tokens:
+        return None
+    if isinstance(tokens, list):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:49]: swap the condition

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -34,9 +44,24 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        freq = {}
+        all_tokens = len(tokens)
+        for token in tokens:
+            if token in freq:
+                freq[token] += 1
+            else:
+                freq[token] = 1
+        for token in freq:
+            freq[token] /= all_tokens
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:58]: can you combine 2 for loops in one?

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -34,9 +44,24 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        return freq
+    return None
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:64]: .

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -50,8 +75,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     In case of corrupt input arguments, None is returned
     """
 
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:78]: swap

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -50,8 +75,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+        freq_dict = calculate_frequencies(tokenize(text))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:79]: what if this func will return None?

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -65,11 +97,22 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     """
 
+    if len(actual) != len(predicted):
+        return None
+    summation = 0
+    n = len(actual)
+    for i in range(0, n):
+        squared_difference = (actual[i] - predicted[i]) ** 2
+        summation = summation + squared_difference
+    mse = summation / n
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:107]: return as is

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -85,6 +128,20 @@ def compare_profiles(
...
+    if not unknown_profile or not profile_to_compare:
+        return None
+    if (not isinstance(unknown_profile['freq'], dict)
+            and not isinstance(profile_to_compare['freq'], dict)):
+        return None
+    if not ('name' in unknown_profile and 'name' in profile_to_compare):
+        return None
+    if not ('freq' in unknown_profile and 'freq' in profile_to_compare):
+        return None
+    all_tokens = set(unknown_profile['freq'].keys()).union(set(profile_to_compare['freq'].keys()))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:140]: can you explain what is going here?

File: lab_1_classify_by_unigrams/settings.json [hunk 10]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 8
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: nice aim! Maybe 10??

File: lab_1_classify_by_unigrams/start.py [hunk 11]
```diff
@@ -3,18 +3,37 @@
...
+    unknown_profile = create_language_profile('unknown', unknown_text)
+
+    print(f'Tokens: {tokens}')
+    print(f'Frequencies: {calculate_frequencies(tokens)}')
+    print(f'Language profile: {create_language_profile("en", en_text)}')
+    print(f'MSE:{calculate_mse}')
+    print(f'Compare profiles:{compare_profiles(en_profile, de_profile)}')
+    print(f'Detected Language: {detect_language(unknown_profile, en_profile, de_profile)}')
+
+    result = "a"
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:35]: make result equals your detected language

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:35]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:49]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:64]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:58]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:78]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:79]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:107]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:140]: If the key 'freq' isnt in unknown_profile and in profile_to_compare None is returned

all_tokens is counting tokens by uniting two sets of keys (which are extracted from the dicts), making another set

unknown_profile is the dict where the key 'freq'  is stored, unknown_profile['freq'] is the needed token and .keys() extracts it

File: lab_1_classify_by_unigrams/start.py (hunk 11 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/start.py:35]: Done

**Review by EIzharikov — COMMENTED** [review]
Nice progress! I believe that you can achieve 10! Be ready for the next review. Good luck!

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -35,6 +43,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if not isinstance(tokens, list) or len(tokens) == 0 or None in tokens:
+        return None
+    freq = {}
+    all_tokens = len(tokens)
+    for token in tokens:
+        if token in freq:
+            freq[token] += 1 / all_tokens
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:52]: can you count ` 1 / all_tokens` earlier?

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -50,6 +69,13 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     In case of corrupt input arguments, None is returned
     """
 
+    if not (isinstance(language, str) and isinstance(text, str)):
+        return None
+    freq_dict = calculate_frequencies(tokenize(text))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:74]: what if `tokenize(text)` will return None?

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -65,6 +91,15 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if len(actual) != len(predicted):
+        return None
+    summation = 0
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:96]: Is `summations` eng word?

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -106,6 +155,27 @@ def detect_language(
...
+        mse_prof_2 = compare_profiles(unknown_profile, profile_2)
+        if mse_prof_1 is not None and mse_prof_2 is not None:
+            prof_1_name = profile_1.get('name')
+            prof_2_name = profile_2.get('name')
+            if isinstance(prof_1_name, str) and isinstance(prof_2_name, str):
+                if mse_prof_1 < mse_prof_2:
+                    return prof_1_name
+                if mse_prof_1 > mse_prof_2:
+                    return prof_2_name
+                return min(prof_1_name, prof_2_name)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:176]: can you explain why are you using `min()` here?

**Review by EIzharikov — COMMENTED** [review]
Nice work, make CI green, fix&answer all my comments and be ready for lab work defense! Good luck!

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:52]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:96]: Yeah… But now I guess summ will be better

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (ShrubberyBush) [reply lab_1_classify_by_unigrams/main.py:176]: If mse_prof_1 = mse_prof_2 then it'll return the profile which name will be the first in the alphabet order

**Pull request closed without merging**